### ISSUE TITLE
Support for IRF file format evolution

### ISSUE THREAD

**Issue by kosack** [gammapy/gammapy#1212]
IRF files will likely evolve in time as new dimensions or keywords are added to keep up with pipeline developments. Therefore it is necessary to think now about how to handle this evolution. 

For IRFs (and really any other support file), it would be nice to make the`read()` and `write()` methods of the IRF classes use reader/write code based on the version number in the file so that the same interface can be used even if the IRF format evolves in time.  That implies also a format version number in the header. 

Perhaps something like this is a simple solution:

```py3
   @classmethod
   def read(cls, filename):
      version = get_irf_version_from_file(filename)
      reader_dispatch = {
          1: cls._read_v1,
          2: cls._read_v2
      }
      return reader_dispatch.get(version)(filename)
```

Where `read_v1()` and `read_v2()` implement reading different format versions.

Or a more general piece of code could be developed to handle this for all file types.

**Comment by cdeil** [comment]
I think this is a good idea.

I'm not sure if we have the versioning in the spec sorted out properly yet. @TarekHC - I know you added the HDU class declarative scheme for different IRFs:
http://gamma-astro-data-formats.readthedocs.io/en/latest/general/hduclass.html
but did we ever agree on versioning, i.e. whether it's per "format" or globally?

It might be easiest to wait for the first format change to occur and to implement it then, or to add an intentional change (like a column rename) just to trigger this use case that CTA will probably have. Otherwise there's a risk that we implement some scheme in Gammapy now, but then find out that it doesn't work well in practice.

There's also implementation details that we have to think through / discuss a bit. E.g. probably it would be nice if the `reader_dispatch` you have as a local variable in `read` is instead a class- or instance-level property, so that one can access it and e.g. create a global list or registry of all supported IRF classes and versions. Another point to discuss would be if `get_irf_version_from_file` is a good idea, or if there should be a larger `get_irf_info` that also figures out the class, i.e. a registry.

So basically: if someone wants to add this, it's welcome of course, but unless it's quick to just implement for you, it might make sense to discuss a bit more on Github or in a call how exactly to do it before spending time to implement.

**Comment by TarekHC** [comment]
Hi @cdeil and @kosack ,

> I'm not sure if we have the versioning in the spec sorted out properly yet. @TarekHC - I know you added the HDU class declarative scheme for different IRFs:
http://gamma-astro-data-formats.readthedocs.io/en/latest/general/hduclass.html
but did we ever agree on versioning, i.e. whether it's per "format" or globally?

I remember we discussed about it, but we never took any decision. 

I can imagine, as all data will be periodically re-analyzed, that new IRF versions will usually go with new productions implementing better parameterizations, etc... So the IRF version could be associated to the "production number" (if such a thing will exist). Then there should also be bugfixes, etc... I don't have a lot of experience with this kind of problems, so anyone with more experience should find a good implementation (although I doubt it is that complicated). If anyone wants to implement something, I could help with adding it to the IRF specs.

I'm actually worried with the fact that there will not be just one IRF reader if we have several science tools... (e.g. ctools/gammapy). That decentralization will just complicate things and be prone to bugs...

**Comment by registerrier** [comment]
In #5015 examples of definition of asymetric IRFs are given. Support for non radially symmetric IRFs in `gammapy.makers` is introduced.

**Comment by registerrier** [comment]
This is an old issue. Several specific ones cover asymetric IRFs or the CTAO data model. Closing this old issue.
